Check internal key range when compacting from last level to penultimate level

cbi42 opened the pull request:
Summary: The test failure in #11909 shows that we may compact keys outside of internal key range of penultimate level input files from last level to penultimate level, which can potentially cause overlapping files in the penultimate level. This PR updates the  `Compaction::WithinPenultimateLevelOutputRange()` to check internal key range instead of user key.

Other fixes:
* skip range del sentinels when deciding output level for tiered compaction

Test plan: 
- existing unit tests
- apply the fix to #11905 and run `./tiered_compaction_test --gtest_filter="*RangeDelsCauseFileEndpointsToOverlap*"`

ajkr commented on db/compaction/compaction.cc:
```diff
@@ -426,33 +467,39 @@ bool Compaction::OverlapPenultimateLevelOutputRange(
   return ucmp->CompareWithoutTimestamp(
-             smallest_key, penultimate_level_largest_user_key_) <= 0 &&
+             smallest_key, penultimate_level_largest_.user_key()) <= 0 &&
          ucmp->CompareWithoutTimestamp(
-             largest_key, penultimate_level_smallest_user_key_) >= 0;
+             largest_key, penultimate_level_smallest_.user_key()) >= 0;
```
If it were me I'd probably just make the comparison here and in `WithinPenultimateLevelOutputRange()` exclusive. But your ways nicer I think.

cbi42 replied:
Thanks, I did not think of that way, it should work too.

ajkr reviewed (approved):
Can a test like TieredCompactionTest.RangeBasedTieredStorageLevel be modified to cover this case? The previous way it was caught seemed like a surprising coincidence

cbi42 commented:
> Can a test like TieredCompactionTest.RangeBasedTieredStorageLevel be modified to cover this case? The previous way it was caught seemed like a surprising coincidence

Yes, I added a repro in `TieredCompactionTest.CheckInternalKeyRange` and updated `TieredCompactionTest.RangeBasedTieredStorageLevel` to include a related test case. I was not sure how to write a unit test for this, the set up in TieredCompactionTest.CheckInternalKeyRange helps 👍 

Thanks for the review!
